Exception during conversion can make the process hang

While performing our QA tests for 0.6.1 on Windows, we tested some success and error cases, to see how our termination logic (see #749 and #772) works for them. Below are the logs for each run:

<table>
  <tr>
    <td>
<details>
  <summary>Successful conversion</summary>
  </br>
  
```
$ .\dev_scripts\dangerzone-cli .\tests\test_docs\sample_bmp.bmp
Converting document to safe PDF
[DEBUG] Marking doc 1648eq as 'converting'
[INFO ] > 'C:\Program Files\Docker\Docker\resources\bin\docker.EXE' run --network none -u dangerzone --security-opt=no-new-privileges:true --cap-drop all --rm -i --name dangerzone-doc-to-pixels-1648eq dangerzone.rocks/dangerzone /usr/bin/python3 -m dangerzone.conversion.doc_to_pixels
[INFO ] [doc 1648eq] 0% Converting page 1/1 to pixels
[INFO ] [doc 1648eq] 49% Converted document to pixels
[INFO ] Conversion output (doc to pixels)
----- DOC TO PIXELS LOG START -----
Converting page 1/1 to pixels
Converted document to pixels
----- DOC TO PIXELS LOG END -----
[WARNING] Failed to kill container dangerzone-doc-to-pixels-1648eq: Command '['C:\\Program Files\\Docker\\Docker\\resources\\bin\\docker.EXE', 'kill', 'dangerzone-doc-to-pixels-1648eq']' returned non-zero exit status 1.
[WARNING] Stdout from the kill command: b''
[WARNING] Stderr from the kill command: b'Error response from daemon: cannot kill container: dangerzone-doc-to-pixels-1648eq: No such container: dangerzone-doc-to-pixels-1648eq\n'
[INFO ] > 'C:\Program Files\Docker\Docker\resources\bin\docker.EXE' run --network none -u dangerzone --security-opt=no-new-privileges:true --cap-drop all --rm -i --name dangerzone-pixels-to-pdf-1648eq -v 'C:\Users\apyrg\AppData\Local\Temp\tmpc_8gg2ys:/safezone:Z' -e OCR=0 -e OCR_LANGUAGE=None dangerzone.rocks/dangerzone /usr/bin/python3 -m dangerzone.conversion.pixels_to_pdf
[INFO ] [doc 1648eq] 50% Converting page 1/1 from pixels to PDF
[INFO ] [doc 1648eq] 100% Safe PDF created
[INFO ] Conversion output: (pixels to PDF)
----- PIXELS TO PDF LOG START -----

----- PIXELS TO PDF LOG END -----
[DEBUG] Marking doc 1648eq as 'safe'
```
 
 </details>
     </td>
  </tr>
    <tr>
    <td>
<details>
  <summary>Failed conversion by the conversion process</summary>
  </br>
  
```
$ .\dev_scripts\dangerzone-cli .\tests\test_docs\sample_bad_pdf.pdf
Converting document to safe PDF
[DEBUG] Marking doc Bd5ntr as 'converting'
[INFO ] > 'C:\Program Files\Docker\Docker\resources\bin\docker.EXE' run --network none -u dangerzone --security-opt=no-new-privileges:true --cap-drop all --rm -i --name dangerzone-doc-to-pixels-Bd5ntr dangerzone.rocks/dangerzone /usr/bin/python3 -m dangerzone.conversion.doc_to_pixels
Exception: The document format is not supported
[ERROR] [doc Bd5ntr] 0% The document format is not supported
[DEBUG] Marking doc Bd5ntr as 'failed'

Failed to convert document(s)
```
 
 </details>
     </td>
  </tr>
    <tr>
    <td>
<details>
  <summary>Failed conversion by the host</summary>
  </br>
  
```
$ .\dev_scripts\dangerzone-cli .\tests\test_docs\sample_bad_max_width.pdf
Converting document to safe PDF
[DEBUG] Marking doc ervQpR as 'converting'
[INFO ] > 'C:\Program Files\Docker\Docker\resources\bin\docker.EXE' run --network none -u dangerzone --security-opt=no-new-privileges:true --cap-drop all --rm -i --name dangerzone-doc-to-pixels-ervQpR dangerzone.rocks/dangerzone /usr/bin/python3 -m dangerzone.conversion.doc_to_pixels
[INFO ] [doc ervQpR] 0% Converting page 1/1 to pixels
[WARNING] Failed to kill container dangerzone-doc-to-pixels-ervQpR: Command '['C:\\Program Files\\Docker\\Docker\\resources\\bin\\docker.EXE', 'kill', 'dangerzone-doc-to-pixels-ervQpR']' returned non-zero exit status 1.
[WARNING] Stdout from the kill command: b''
[WARNING] Stderr from the kill command: b'Error response from daemon: cannot kill container: dangerzone-doc-to-pixels-ervQpR: container 07a0b4524bc8f43f75ea1a3fca278fcc13598bff910f09216a3dffc176c24ddd is not running\n'
[WARNING] Conversion process did not terminate gracefully after 15 seconds. Killing it forcefully...
[ERROR] [doc ervQpR] 0% A page exceeded the maximum width.
[DEBUG] Marking doc ervQpR as 'failed'

Failed to convert document(s)
```
 
 </details>
     </td>
  </tr>
</table>

Two main things are of note here:

1. In some error cases, our termination logic reports that the container does not exist, whereas the conversion process (`docker run ...`) still exists.
2. Specifically for the failed conversion case, when that happens in the host, the conversion process seems to be stuck indefinitely, and we have to kill it.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,9 +15,13 @@ since 0.4.1, and this project adheres to [Semantic Versioning](https://semver.or
 
 ### Fixed
 
+- Fix "Timeout after 3 seconds" errors ([issue #749](https://github.com/freedomofpress/dangerzone/issues/749))
 - Make OCR work again in Qubes Fedora 38 templates ([issue #737](https://github.com/freedomofpress/dangerzone/issues/737))
 - Make .svg / .bmp files selectable when browsing files via the Dangerzone GUI ([#722](https://github.com/freedomofpress/dangerzone/pull/722))
-- Linux: Allow opening multiple files at once, when selecting them from the user's file manager ([issue #797](https://github.com/freedomofpress/dangerzone/issues/797))
+- Linux: Show the proper application name and icon for Dangerzone, in the user's window manager, thanks to [@naglis](https://github.com/naglis) ([issue #402](https://github.com/freedomofpress/dangerzone/issues/402))
+- Linux: Allow opening multiple files at once, when selecting them from the user's file manager, thanks to [@naglis](https://github.com/naglis) ([issue #797](https://github.com/freedomofpress/dangerzone/issues/797))
+- Linux: Do not include Dangerzone in the list of available PDF viewers, thanks to [@naglis](https://github.com/naglis) ([issue #790](https://github.com/freedomofpress/dangerzone/issues/790))
+- Linux: Handle filenames with invalid Unicode characters in the Dangerzone CLI, thanks to [@naglis](https://github.com/naglis) ([issue #768](https://github.com/freedomofpress/dangerzone/issues/768))
 
 ### Changed
 

--- a/RELEASE.md
+++ b/RELEASE.md
@@ -15,7 +15,7 @@ Before making a release, all of these should be complete:
 - [ ] Update the "Version" field in `install/linux/dangerzone.spec`
 - [ ] Update screenshot in `README.md`, if necessary
 - [ ] CHANGELOG.md should be updated to include a list of all major changes since the last release
-- [ ] Create a PGP-signed git tag for the version, e.g. for dangerzone `v0.1.0`, the is created with:
+- [ ] Create a PGP-signed git tag for the version, e.g., for dangerzone `v0.1.0`:
 
   ```
   git tag -s v0.1.0
@@ -176,7 +176,7 @@ dangerzone.rocks/dangerzone  latest      <different ID>  <newer date>  <differen
 
 #### 4. Dangerzone successfully installs the container image
 
-_(Linux)_
+_(Only for Linux)_
 
 Remove the Dangerzone container image from Docker/Podman. Then run Dangerzone.
 Dangerzone should install the container image successfully.

--- a/dangerzone/args.py
+++ b/dangerzone/args.py
@@ -1,5 +1,6 @@
 import functools
 import os
+import sys
 from typing import List, Optional, Tuple
 
 import click
@@ -86,7 +87,7 @@ def check_suspicious_options(args: List[str]) -> None:
             f" current working directory: {filenames_str}"
         )
         click.echo(msg)
-        exit(1)
+        sys.exit(1)
 
 
 def override_parser_and_check_suspicious_options(click_main: click.Command) -> None:

--- a/dangerzone/cli.py
+++ b/dangerzone/cli.py
@@ -67,7 +67,7 @@ def cli_main(
         dangerzone.add_document_from_filename(filenames[0], output_filename, archive)
     elif len(filenames) > 1 and output_filename:
         click.echo("--output-filename can only be used with one input file.")
-        exit(1)
+        sys.exit(1)
     else:
         for filename in filenames:
             dangerzone.add_document_from_filename(filename, archive=archive)
@@ -83,7 +83,7 @@ def cli_main(
             click.echo("Invalid OCR language code. Valid language codes:")
             for lang in dangerzone.ocr_languages:
                 click.echo(f"{dangerzone.ocr_languages[lang]}: {lang}")
-            exit(1)
+            sys.exit(1)
 
     # Ensure container is installed
     dangerzone.isolation_provider.install()

--- a/dangerzone/document.py
+++ b/dangerzone/document.py
@@ -145,6 +145,8 @@ class Document:
         new_file_path = archive_dir / old_file_path.name
         log.debug(f"Archiving doc {self.id} to {new_file_path}")
         Path.mkdir(archive_dir, exist_ok=True)
+        # On Windows, moving the file will fail if it already exists.
+        new_file_path.unlink(missing_ok=True)
         old_file_path.rename(new_file_path)
 
     @property

--- a/dangerzone/errors.py
+++ b/dangerzone/errors.py
@@ -107,6 +107,6 @@ def handle_document_errors(func: F) -> F:
                 msg = "An exception occured while validating a document"
                 log.exception(msg)
             click.echo(str(e))
-            exit(1)
+            sys.exit(1)
 
     return cast(F, wrapper)

--- a/dangerzone/gui/logic.py
+++ b/dangerzone/gui/logic.py
@@ -111,7 +111,7 @@ class DangerzoneGui(DangerzoneCore):
                             desktop_entry = DesktopEntry(full_filename)
                             if (
                                 "application/pdf" in desktop_entry.getMimeTypes()
-                                and desktop_entry.getName() != "dangerzone"
+                                and "dangerzone" not in desktop_entry.getName().lower()
                             ):
                                 pdf_viewers[desktop_entry.getName()] = (
                                     desktop_entry.getExec()

--- a/dangerzone/isolation_provider/base.py
+++ b/dangerzone/isolation_provider/base.py
@@ -45,9 +45,9 @@ def read_int(f: IO[bytes]) -> int:
 
 
 def read_debug_text(f: IO[bytes], size: int) -> str:
-    """Read arbitrarily long text (for debug purposes)"""
+    """Read arbitrarily long text (for debug purposes), and sanitize it."""
     untrusted_text = f.read(size).decode("ascii", errors="replace")
-    return replace_control_chars(untrusted_text)
+    return replace_control_chars(untrusted_text, keep_newlines=True)
 
 
 class IsolationProvider(ABC):

--- a/dangerzone/isolation_provider/base.py
+++ b/dangerzone/isolation_provider/base.py
@@ -146,10 +146,13 @@ class IsolationProvider(ABC):
 
         if getattr(sys, "dangerzone_dev", False):
             assert p.stderr
-            untrusted_log = read_debug_text(p.stderr, MAX_CONVERSION_LOG_CHARS)
+            debug_log = read_debug_text(p.stderr, MAX_CONVERSION_LOG_CHARS)
             p.stderr.close()
             log.info(
-                f"Conversion output (doc to pixels)\n{self.sanitize_conversion_str(untrusted_log)}"
+                "Conversion output (doc to pixels)\n"
+                f"{DOC_TO_PIXELS_LOG_START}\n"
+                f"{debug_log}"  # no need for an extra newline here
+                f"{DOC_TO_PIXELS_LOG_END}"
             )
 
     @abstractmethod
@@ -195,14 +198,6 @@ class IsolationProvider(ABC):
     @abstractmethod
     def get_max_parallel_conversions(self) -> int:
         pass
-
-    def sanitize_conversion_str(self, untrusted_conversion_str: str) -> str:
-        conversion_string = replace_control_chars(untrusted_conversion_str)
-
-        # Add armor (gpg-style)
-        armor_start = f"{DOC_TO_PIXELS_LOG_START}\n"
-        armor_end = DOC_TO_PIXELS_LOG_END
-        return armor_start + conversion_string + armor_end
 
     @abstractmethod
     def start_doc_to_pixels_proc(self, document: Document) -> subprocess.Popen:

--- a/dangerzone/isolation_provider/container.py
+++ b/dangerzone/isolation_provider/container.py
@@ -228,18 +228,16 @@ class Container(IsolationProvider):
         container_runtime = self.get_runtime()
         cmd = [container_runtime, "kill", name]
         try:
+            # We do not check the exit code of the process here, since the container may
+            # have stopped right before invoking this command. In that case, the
+            # command's output will contain some error messages, so we capture them in
+            # order to silence them.
             subprocess.run(
-                cmd,
-                capture_output=True,
-                check=True,
-                startupinfo=get_subprocess_startupinfo(),
+                cmd, capture_output=True, startupinfo=get_subprocess_startupinfo()
             )
-        except subprocess.CalledProcessError as e:
-            log.warning(f"Failed to kill container {name}: {str(e)}")
-            log.warning(f"Output from the kill command: {e.output}")
         except Exception as e:
             log.exception(
-                f"Unexpected error occurred while killing container {name}: {str(e)}"
+                f"Unexpected error occurred while killing container '{name}': {str(e)}"
             )
 
     def pixels_to_pdf(
@@ -303,7 +301,33 @@ class Container(IsolationProvider):
     def terminate_doc_to_pixels_proc(
         self, document: Document, p: subprocess.Popen
     ) -> None:
+        # There are two steps to gracefully terminate a conversion process:
+        # 1. Kill the container, and check that it has exited.
+        # 2. Gracefully terminate the conversion process, in case it's stuck on I/O
+        #
+        # See also https://github.com/freedomofpress/dangerzone/issues/791
         self.kill_container(self.doc_to_pixels_container_name(document))
+        p.terminate()
+
+    def ensure_stop_doc_to_pixels_proc(  # type: ignore [no-untyped-def]
+        self, document: Document, *args, **kwargs
+    ) -> None:
+        super().ensure_stop_doc_to_pixels_proc(document, *args, **kwargs)
+
+        # Check if the container no longer exists, either because we successfully killed
+        # it, or because it exited on its own. We operate under the assumption that
+        # after a podman kill / docker kill invocation, this will likely be the case,
+        # else the container runtime (Docker/Podman) has experienced a problem, and we
+        # should report it.
+        container_runtime = self.get_runtime()
+        name = self.doc_to_pixels_container_name(document)
+        all_containers = subprocess.run(
+            [container_runtime, "ps", "-a"],
+            capture_output=True,
+            startupinfo=get_subprocess_startupinfo(),
+        )
+        if name in all_containers.stdout.decode():
+            log.warning(f"Container '{name}' did not stop gracefully")
 
     def get_max_parallel_conversions(self) -> int:
         # FIXME hardcoded 1 until length conversions are better handled

--- a/dangerzone/isolation_provider/container.py
+++ b/dangerzone/isolation_provider/container.py
@@ -228,7 +228,12 @@ class Container(IsolationProvider):
         container_runtime = self.get_runtime()
         cmd = [container_runtime, "kill", name]
         try:
-            subprocess.run(cmd, capture_output=True, check=True)
+            subprocess.run(
+                cmd,
+                capture_output=True,
+                check=True,
+                startupinfo=get_subprocess_startupinfo(),
+            )
         except subprocess.CalledProcessError as e:
             log.warning(f"Failed to kill container {name}: {str(e)}")
             log.warning(f"Output from the kill command: {e.output}")

--- a/dangerzone/util.py
+++ b/dangerzone/util.py
@@ -66,11 +66,14 @@ def get_subprocess_startupinfo():  # type: ignore [no-untyped-def]
         return None
 
 
-def replace_control_chars(untrusted_str: str) -> str:
+def replace_control_chars(untrusted_str: str, keep_newlines: bool = False) -> str:
     """Remove control characters from string. Protects a terminal emulator
     from obscure control characters.
 
     Control characters are replaced by � U+FFFD Replacement Character.
+
+    If a user wants to keep the newline character (e.g., because they are sanitizing a
+    multi-line text), they must pass `keep_newlines=True`.
     """
 
     def is_safe(chr: str) -> bool:
@@ -90,5 +93,8 @@ def replace_control_chars(untrusted_str: str) -> str:
 
     sanitized_str = ""
     for char in untrusted_str:
-        sanitized_str += char if is_safe(char) else "�"
+        if (keep_newlines and char == "\n") or is_safe(char):
+            sanitized_str += char
+        else:
+            sanitized_str += "�"
     return sanitized_str

--- a/dev_scripts/env.py
+++ b/dev_scripts/env.py
@@ -608,6 +608,7 @@ class Env:
     ):
         """Build a Linux environment and install Dangerzone in it."""
         build_dir = distro_build(self.distro, self.version)
+        os.makedirs(build_dir, exist_ok=True)
         version = dz_version()
         if self.distro == "fedora":
             install_deps = DOCKERFILE_BUILD_FEDORA_DEPS
@@ -671,8 +672,6 @@ class Env:
         if show_dockerfile:
             print(dockerfile)
             return
-
-        os.makedirs(build_dir, exist_ok=True)
 
         # Populate the build context.
         shutil.copy(package_src, package_dst)

--- a/dev_scripts/qa.py
+++ b/dev_scripts/qa.py
@@ -120,7 +120,7 @@ dangerzone.rocks/dangerzone  latest      <different ID>  <newer date>  <differen
 
 #### 4. Dangerzone successfully installs the container image
 
-_(Linux)_
+_(Only for Linux)_
 
 Remove the Dangerzone container image from Docker/Podman. Then run Dangerzone.
 Dangerzone should install the container image successfully.

--- a/dev_scripts/qa.py
+++ b/dev_scripts/qa.py
@@ -430,7 +430,7 @@ class Reference:
                 " to date with the respective doc section, and then update the cached"
                 " section in this file."
             )
-            exit(1)
+            sys.exit(1)
 
     def find_section_text(self, md_text):
         """Find a section's content in a provided Markdown string."""
@@ -978,7 +978,7 @@ def parse_args():
 
     if not args.check_refs and not args.platform:
         parser.print_help(sys.stderr)
-        exit(1)
+        sys.exit(1)
 
     return args
 

--- a/install/macos/build-app.py
+++ b/install/macos/build-app.py
@@ -7,6 +7,7 @@ import itertools
 import os
 import shutil
 import subprocess
+import sys
 
 root = os.path.dirname(
     os.path.dirname(
@@ -69,7 +70,7 @@ def sign_app_bundle(build_path, dist_path, app_path):
     # Detect if create-dmg is installed
     if not os.path.exists(app_path):
         print(f"ERROR: Dangerzone.app not found in {app_path}.")
-        exit(1)
+        sys.exit(1)
 
     dmg_path = os.path.join(dist_path, "Dangerzone.dmg")
     icon_path = os.path.join(root, "install", "macos", "dangerzone.icns")

--- a/tests/__init__.py
+++ b/tests/__init__.py
@@ -123,15 +123,23 @@ def uncommon_text() -> str:
 def uncommon_filename(uncommon_text: str) -> str:
     """Craft a filename with Unicode characters that are considered not common.
 
-    We reuse the same uncommon string as above, with a small exception for macOS.
-    Because the APFS filesystem in macOS accepts only UTF-8 encoded strings [1], we
-    cannot create a filename with invalid Unicode characters. So, in order to test the
-    rest of the corner cases, we replace U+DCF0 with an empty string.
+    We reuse the same uncommon string as above, with a small exception for macOS and
+    Windows.
+
+    Because the NTFS filesystem in Windows and APFS filesystem in macOS accept only
+    UTF-8 encoded strings [1], we cannot create a filename with invalid Unicode
+    characters. So, in order to test the rest of the corner cases, we replace U+DCF0
+    with an empty string.
+
+    Windows has the extra restriction that it cannot have escape characters in
+    filenames, so we replace the ASCII Escape character (\033 / U+001B) as well.
 
     [1]: https://en.wikipedia.org/wiki/Filename#Comparison_of_filename_limitations
     """
     if platform.system() == "Darwin":
         uncommon_text = uncommon_text.replace("\udcf0", "")
+    elif platform.system() == "Windows":
+        uncommon_text = uncommon_text.replace("\udcf0", "").replace("\033", "")
     return uncommon_text + ".pdf"
 
 

--- a/tests/test_document.py
+++ b/tests/test_document.py
@@ -97,24 +97,28 @@ def test_archive_unwriteable_dir(tmp_path: Path) -> None:
 
 
 def test_archive(mocker: MagicMock, tmp_path: Path) -> None:
-    test_string = "original file"
     original_doc_path = str(tmp_path / "doc.pdf")
     archived_doc_path = str(tmp_path / ARCHIVE_SUBDIR / "doc.pdf")
 
-    # write some content for later verifying content integrity
-    with open(original_doc_path, "w") as f:
-        f.write(test_string)
+    # Perform the archival operation two times: one with no archive dir, and one with an
+    # archive dir.
+    test_strings = ["original file 1", "original file 2"]
+    for test_string in test_strings:
+        # write some content for later verifying content integrity
+        with open(original_doc_path, "w") as f:
+            f.write(test_string)
 
-    d = Document(original_doc_path, archive=True)
-    d.archive()
+        # archive the document
+        d = Document(original_doc_path, archive=True)
+        d.archive()
 
-    # original document has been moved to unsafe/doc.pdf
-    assert not os.path.exists(original_doc_path)
-    assert os.path.exists(archived_doc_path)
+        # original document has been moved to unsafe/doc.pdf
+        assert not os.path.exists(original_doc_path)
+        assert os.path.exists(archived_doc_path)
 
-    # make sure it is the original file by comparing its content
-    with open(archived_doc_path) as f:
-        assert f.read() == test_string
+        # make sure it is the proper file by comparing its content
+        with open(archived_doc_path) as f:
+            assert f.read() == test_string
 
 
 def test_set_output_dir(sample_pdf: str, tmp_path: Path) -> None:

--- a/tests/test_large_set.py
+++ b/tests/test_large_set.py
@@ -1,6 +1,7 @@
 import os
 import re
 import subprocess
+import sys
 import time
 from pathlib import Path
 from typing import List
@@ -23,7 +24,7 @@ FORMATS_REGEX = (
 def ensure_test_data_exists() -> None:
     if len(os.listdir(test_docs_repo_dir)) == 0:
         print("Test data repository it empty. Skipping large tests.")
-        exit(1)
+        sys.exit(1)
 
 
 def get_test_docs(min_size: int, max_size: int) -> List[Path]:

--- a/tests/test_util.py
+++ b/tests/test_util.py
@@ -30,3 +30,8 @@ def test_replace_control_chars(uncommon_text: str, sanitized_text: str) -> None:
     assert util.replace_control_chars(uncommon_text) == sanitized_text
     assert util.replace_control_chars("normal text") == "normal text"
     assert util.replace_control_chars("") == ""
+    assert util.replace_control_chars("multi-line\ntext") == "multi-line�text"
+    assert (
+        util.replace_control_chars("multi-line\ntext", keep_newlines=True)
+        == "multi-line\ntext"
+    )

